qiskit/ignis/verification/randomized_benchmarking/dihedral.py

```diff
@@ -333,53 +338,53 @@ def __str__(self):
 
 class CNOTDihedral():
     """CNOT-dihedral Object Class.
-    The CNOT-dihedral group on n qubits is generated by the gates
+    The CNOT-dihedral group on num_qubits qubits is generated by the gates
     CNOT, T and X."""
 
-    def __init__(self, n_qubits):
-        # Construct the identity element on n_qubits qubits.
-        self.n_qubits = n_qubits
+    def __init__(self, num_qubits):
+        # Construct the identity element on num_qubits qubits.
+        self.num_qubits = num_qubits
         # phase polynomial
-        self.poly = SpecialPolynomial(n_qubits)
+        self.poly = SpecialPolynomial(num_qubits)
         # n x n invertible matrix over Z_2
-        self.linear = [[int(r == c) for c in range(n_qubits)]
-                       for r in range(n_qubits)]
+        self.linear = [[int(r == c) for c in range(num_qubits)]
```

Comment:
These would probably be much faster in functions that use them if they were Numpy `boolean` or `int` arrays.

I think this function is making an identity matrix so could be: `np.eye(num_qubits, dtype=np.bool)` or `np.eye(num_qubits, dtype=np.int8)`

Reply:
> 
> 
> The overall structure looks good, but there are a lot of inefficiencies in the Dihedral group method using nested python for loops, if these were rewritten to use Numpy array and array methods (dot etc), it should give performance improvements.

I agree with your comment. Indeed, the current CNOTDihedral class has many for loops, and does not use numpy arrays.
However, it this PR I haven't changed the code of the existing CNOTDihedral class, and only added some extra functionality. 
I suggest to merge this PR (since there are other PRs that depend on it), and then I will open another PR for rewriting the code to improve the performance of the CNOTDihedral class.